#### Extracción y exploración

In [2]:
import pandas as pd
import ast
import tools
import warnings
warnings.filterwarnings("ignore")


ruta_items = "australian_user_reviews.json"

# Se lee de cada línea del dataset
filas_items = []
with open(ruta_items, "r", encoding="utf-8") as f:
    for line in f.readlines():
        filas_items.append(ast.literal_eval(line))

# Se convierte en dataframe
df_user_reviews = pd.DataFrame(filas_items)
df_user_reviews

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1                js41637               http://steamcommunity.com/id/js41637   
2              evcentric             http://steamcommunity.com/id/evcentric   
3                  doctr                 http://steamcommunity.com/id/doctr   
4              maplemage             http://steamcommunity.com/id/maplemage   
...                  ...                                                ...   
25794  76561198306599751  http://steamcommunity.com/profiles/76561198306...   
25795           Ghoustik              http://steamcommunity.com/id/Ghoustik   
25796  76561198310819422  http://steamcommunity.com/profiles/76561198310...   
25797  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
25798        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                                 reviews  
0      [{'funny': '', 'posted': 'Posted November 5, 2...  
1      [{'funny': '', 'posted': 'Posted June 24, 2014...  
2      [{'funny': '', 'posted': 'Posted February 3.',...  
3      [{'funny': '', 'posted': 'Posted October 14, 2...  
4      [{'funny': '3 people found this review funny',...  
...                                                  ...  
25794  [{'funny': '', 'posted': 'Posted May 31.', 'la...  
25795  [{'funny': '', 'posted': 'Posted June 17.', 'l...  
25796  [{'funny': '1 person found this review funny',...  
25797  [{'funny': '', 'posted': 'Posted July 21.', 'l...  
25798  [{'funny': '1 person found this review funny',...  

[25799 rows x 3 columns]

In [3]:
#Verificamos el tipo de datos que contiene
tools.ver_tipo_datos(df_user_reviews)

nombre_campo        tipo_datos  no_nulos_%  nulos_%  nulos
0      user_id   [<class 'str'>]       100.0      0.0      0
1     user_url   [<class 'str'>]       100.0      0.0      0
2      reviews  [<class 'list'>]       100.0      0.0      0

Observamos que no hay datos nulos, pero la columna reviews es una lista, lo más seguro es que tengamos que desanidarla. 

In [24]:
filas_duplicadas = tools.duplicados(df_user_reviews, "user_id")
filas_duplicadas 

user_id  \
12888               05041129   
5250                05041129   
3133   111222333444555666888   
3134   111222333444555666888   
4139                   29123   
...                      ...   
2721          xXAussieRockXx   
2680             yolofaceguy   
17916            yolofaceguy   
5855               zeroblade   
13975              zeroblade   

                                                user_url  \
12888              http://steamcommunity.com/id/05041129   
5250               http://steamcommunity.com/id/05041129   
3133   http://steamcommunity.com/id/11122233344455566...   
3134   http://steamcommunity.com/id/11122233344455566...   
4139                  http://steamcommunity.com/id/29123   
...                                                  ...   
2721         http://steamcommunity.com/id/xXAussieRockXx   
2680            http://steamcommunity.com/id/yolofaceguy   
17916           http://steamcommunity.com/id/yolofaceguy   
5855              http://steamcommunity.com/id/zeroblade   
13975             http://steamcommunity.com/id/zeroblade   

                                                 reviews  
12888  [{'funny': '', 'posted': 'Posted May 18, 2015....  
5250   [{'funny': '', 'posted': 'Posted May 18, 2015....  
3133   [{'funny': '', 'posted': 'Posted December 22, ...  
3134   [{'funny': '', 'posted': 'Posted December 22, ...  
4139   [{'funny': '', 'posted': 'Posted March 26.', '...  
...                                                  ...  
2721   [{'funny': '', 'posted': 'Posted July 17, 2015...  
2680   [{'funny': '', 'posted': 'Posted October 31, 2...  
17916  [{'funny': '', 'posted': 'Posted October 31, 2...  
5855   [{'funny': '', 'posted': 'Posted November 30, ...  
13975  [{'funny': '', 'posted': 'Posted November 30, ...  

[623 rows x 3 columns]

Podemos apreciar que hay 623 filas repetidas para la columna "user_id". Vamos a revisar si los datos en "reviews" dentro de la lista son iguales o contiene comentarios diferentes.   

In [29]:
user_reviews = filas_duplicadas[filas_duplicadas["user_id"] == "yolofaceguy"]["reviews"]

for i in user_reviews:
    for r in i:
        print(r["review"])
    print("")




from the creaters of the walking dead, i present to you, the wolf among us. a twisted an unhappy place where crimes are made in the town of the fables. Ecperience one of the most mind-bending, jaw-dropping twists you will see in all of gaming history...well, some of gaming history. SPOILERS: i really liked how bigby turns into his true form to fight all the bloody marys... its just like when neo fought all the agent smiths in the matrix. what i didnt really like is when i chose to lock the crooked man up it glitches my game and i start from the near begining where bigby fought the crooked man's crew, exept the guy who runs a strip club wasnt there. so i was really confused. but anyway i really recommend this game if you want twists and ♥♥♥♥ed up scenes.
this game is awesome,this game is ♥♥♥♥ed up and this game is so sad and depressing. if you want these types of games i would strongly reccomend youto buy it. its worth your money

from the creaters of the walking dead, i present to you,

Podemos ver que los comentarios son exactamente iguales. Por lo tanto, decidimos de quedarnos con el primero y eliminar el resto. 

In [30]:
df_user_reviews = df_user_reviews.drop_duplicates(subset="user_id", keep="first")
tools.duplicados(df_user_reviews, "user_id")

'No hay duplicados'

Analizaremos la columna "reviews" para entender más a fondo la estructura de los datos.

In [31]:
df_user_reviews["reviews"][3]

[{'funny': '',
  'posted': 'Posted October 14, 2013.',
  'last_edited': '',
  'item_id': '250320',
  'helpful': '2 of 2 people (100%) found this review helpful',
  'recommend': True,
  'review': 'This game... is so fun. The fight sequences have been improved from walking dead. It also includes more of a Sam and Max puzzle solving (some of it in the first episode) and walking dead. The game also gets even more better if you have read the Fables comic books, which are without a doubt, very good. The music is also superb and fit the scenarios very well.'},
 {'funny': '',
  'posted': 'Posted July 28, 2012.',
  'last_edited': '',
  'item_id': '20920',
  'helpful': '1 of 1 people (100%) found this review helpful',
  'recommend': True,
  'review': "Really Really Really Great Game, very good story, im in chapter 1 atm and i think its great. You get a really early link with characters. No need to play the first game, its nicely wrapped up for you in a five minute video. FYI beware of the sex sc

Este dataFrame contiene 3 columnas y 25799 filas, sin valores nulos. Las columnas son:

* user_id: identificador único para el usuario.
* user_url: la url del perfil del usuario en streamcommunity.
* reviews: una lista de diccionarios. Para cada usuario se tiene uno o más            diccionario con el review. Cada diccionario contiene:
    * funny: indica si alguien puso emoticón de gracioso al review.
    * posted: fecha de posteo del review en formato "Posted April 21, 2011".
    * last_edited: fecha de la última edición.
    * item_id: identificador único del item, es decir, del juego.
    * helpful: estadística donde otros usuarios indican si fue útil la información.
    * recommend: booleano que indica si el usuario recomienda o no el juego.
    * review: sentencia string con los comentarios sobre el juego.

La columna "reviews" se presenta anidada, siendo una lista con uno o más diccionarios como elementos. Buscaremos generar una columna por cada diccionario para posteriormente hacer un registro por cada diccionario.



In [32]:
df_desanidado = pd.json_normalize(df_user_reviews["reviews"])
df_desanidado

0  \
0      {'funny': '', 'posted': 'Posted November 5, 20...   
1      {'funny': '', 'posted': 'Posted June 24, 2014....   
2      {'funny': '', 'posted': 'Posted February 3.', ...   
3      {'funny': '', 'posted': 'Posted October 14, 20...   
4      {'funny': '3 people found this review funny', ...   
...                                                  ...   
25480  {'funny': '', 'posted': 'Posted May 31.', 'las...   
25481  {'funny': '', 'posted': 'Posted June 17.', 'la...   
25482  {'funny': '1 person found this review funny', ...   
25483  {'funny': '', 'posted': 'Posted July 21.', 'la...   
25484  {'funny': '1 person found this review funny', ...   

                                                       1  \
0      {'funny': '', 'posted': 'Posted July 15, 2011....   
1      {'funny': '', 'posted': 'Posted September 8, 2...   
2      {'funny': '', 'posted': 'Posted December 4, 20...   
3      {'funny': '', 'posted': 'Posted July 28, 2012....   
4      {'funny': '1 person found this review funny', ...   
...                                                  ...   
25480                                               None   
25481                                               None   
25482                                               None   
25483  {'funny': '', 'posted': 'Posted July 10.', 'la...   
25484  {'funny': '', 'posted': 'Posted July 20.', 'la...   

                                                       2  \
0      {'funny': '', 'posted': 'Posted April 21, 2011...   
1      {'funny': '', 'posted': 'Posted November 29, 2...   
2      {'funny': '', 'posted': 'Posted November 3, 20...   
3      {'funny': '', 'posted': 'Posted June 2, 2012.'...   
4      {'funny': '2 people found this review funny', ...   
...                                                  ...   
25480                                               None   
25481                                               None   
25482                                               None   
25483  {'funny': '', 'posted': 'Posted July 10.', 'la...   
25484  {'funny': '', 'posted': 'Posted July 2.', 'las...   

                                                       3  \
0                                                   None   
1                                                   None   
2      {'funny': '', 'posted': 'Posted October 15, 20...   
3      {'funny': '', 'posted': 'Posted June 29, 2014....   
4      {'funny': '', 'posted': 'Posted July 11, 2013....   
...                                                  ...   
25480                                               None   
25481                                               None   
25482                                               None   
25483  {'funny': '', 'posted': 'Posted July 8.', 'las...   
25484                                               None   

                                                       4  \
0                                                   None   
1                                                   None   
2      {'funny': '', 'posted': 'Posted October 15, 20...   
3      {'funny': '', 'posted': 'Posted November 22, 2...   
4                                                   None   
...                                                  ...   
25480                                               None   
25481                                               None   
25482                                               None   
25483                                               None   
25484                                               None   

                                                       5     6     7     8  \
0                                                   None  None  None  None   
1                                                   None  None  None  None   
2      {'funny': '', 'posted': 'Posted October 15, 20...  None  None  None   
3      {'funny': '', 'posted': 'Posted February 23, 2...  None  None  None   
4                                                   None  No

Observamos que al realizar esta transformación, perdemos el "user_id" y "user_url" al que pertenece cada diccionario, pero aún mantiene la misma posición. Para volver a tener la trazabilidad del usuario, lo concatenaremos con el dataframe anterior.

In [33]:
df_desanidado = pd.concat([df_user_reviews[["user_id", "user_url"]], df_desanidado], axis=1)
df_desanidado


user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1                js41637               http://steamcommunity.com/id/js41637   
2              evcentric             http://steamcommunity.com/id/evcentric   
3                  doctr                 http://steamcommunity.com/id/doctr   
4              maplemage             http://steamcommunity.com/id/maplemage   
...                  ...                                                ...   
17819                NaN                                                NaN   
17916                NaN                                                NaN   
18028                NaN                                                NaN   
18234                NaN                                                NaN   
18309                NaN                                                NaN   

                                                       0  \
0      {'funny': '', 'posted': 'Posted November 5, 20...   
1      {'funny': '', 'posted': 'Posted June 24, 2014....   
2      {'funny': '', 'posted': 'Posted February 3.', ...   
3      {'funny': '', 'posted': 'Posted October 14, 20...   
4      {'funny': '3 people found this review funny', ...   
...                                                  ...   
17819  {'funny': '', 'posted': 'Posted July 12, 2014....   
17916  {'funny': '', 'posted': 'Posted December 27, 2...   
18028  {'funny': '', 'posted': 'Posted September 22, ...   
18234  {'funny': '', 'posted': 'Posted July 7, 2014.'...   
18309  {'funny': '', 'posted': 'Posted January 27.', ...   

                                                       1  \
0      {'funny': '', 'posted': 'Posted July 15, 2011....   
1      {'funny': '', 'posted': 'Posted September 8, 2...   
2      {'funny': '', 'posted': 'Posted December 4, 20...   
3      {'funny': '', 'posted': 'Posted July 28, 2012....   
4      {'funny': '1 person found this review funny', ...   
...                                                  ...   
17819                                               None   
17916  {'funny': '', 'posted': 'Posted September 2, 2...   
18028                                               None   
18234                                               None   
18309  {'funny': '', 'posted': 'Posted January 27.', ...   

                                                       2  \
0      {'funny': '', 'posted': 'Posted April 21, 2011...   
1      {'funny': '', 'posted': 'Posted November 29, 2...   
2      {'funny': '', 'posted': 'Posted November 3, 20...   
3      {'funny': '', 'posted': 'Posted June 2, 2012.'...   
4      {'funny': '2 people found this review funny', ...   
...                                                  ...   
17819                                               None   
17916  {'funny': '', 'posted': 'Posted August 23, 201...   
18028                                               None   
18234                                               None   
18309  {'funny': '', 'posted': 'Posted January 27.', ...   

                                                       3  \
0                                                   None   
1                                                   None   
2      {'funny': '', 'posted': 'Posted October 15, 20...   
3      {'funny': '', 'posted': 'Posted June 29, 2014....   
4      {'funny': '', 'posted': 'Posted July 11, 2013....   
...                                                  ...   
17819                                               None   
17916                                               None   
18028                                               None   
18234                                               None   
18309  {'funny': '', 'posted': 'Posted January 19.', ...   

                                                       4  \
0                                                   None   
1                                                   None   
2   

Al tener los diccionarios por columnas, generaremos un registro por cada diccionario con el usuario que genera dicha información.

In [34]:
df_desanidado = pd.melt(df_desanidado, id_vars=["user_id", "user_url"], 
                       value_vars=list(range(9)),
                       value_name="reviews")
df_desanidado.head()


user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

  variable                                            reviews  
0        0  {'funny': '', 'posted': 'Posted November 5, 20...  
1        0  {'funny': '', 'posted': 'Posted June 24, 2014....  
2        0  {'funny': '', 'posted': 'Posted February 3.', ...  
3        0  {'funny': '', 'posted': 'Posted October 14, 20...  
4        0  {'funny': '3 people found this review funny', ...

Al hacer lo últimos, observamos que se presentan registros None. Esto ocurre porque hay usuarios que hicieron más reviews que otros. 
En este ejemplo se puede ver este caso:

In [35]:
df_desanidado[df_desanidado["user_id"]=="76561197970982479"]

user_id                                           user_url  \
0       76561197970982479  http://steamcommunity.com/profiles/76561197970...   
25799   76561197970982479  http://steamcommunity.com/profiles/76561197970...   
51598   76561197970982479  http://steamcommunity.com/profiles/76561197970...   
77397   76561197970982479  http://steamcommunity.com/profiles/76561197970...   
103196  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
128995  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
154794  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
180593  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
206392  76561197970982479  http://steamcommunity.com/profiles/76561197970...   

       variable                                            reviews  
0             0  {'funny': '', 'posted': 'Posted November 5, 20...  
25799         1  {'funny': '', 'posted': 'Posted July 15, 2011....  
51598         2  {'funny': '', 'posted': 'Posted April 21, 2011...  
77397         3                                               None  
103196        4                                               None  
128995        5                                               None  
154794        6                                               None  
180593        7                                               None  
206392        8                                               None

Eliminamos los registros que tienen None en "reviews".

In [36]:
df_desanidado = df_desanidado.dropna()
# Verificamos que solo queden el 'user_id' con la cantidad de diccionarios que le corresponde
df_desanidado[df_desanidado["user_id"]=="76561197970982479"]

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
25799  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
51598  76561197970982479  http://steamcommunity.com/profiles/76561197970...   

      variable                                            reviews  
0            0  {'funny': '', 'posted': 'Posted November 5, 20...  
25799        1  {'funny': '', 'posted': 'Posted July 15, 2011....  
51598        2  {'funny': '', 'posted': 'Posted April 21, 2011...

En este punto, ya podemos convertir cada diccionario en columna.

In [37]:
# Separamos por columnas cada una de las claves de "reviews"
df_user_reviews = df_desanidado["reviews"].apply(pd.Series, dtype="object")
df_user_reviews = df_user_reviews.add_prefix("reviews_")
df_user_reviews.head()

reviews_funny            reviews_posted  \
0                                    Posted November 5, 2011.   
1                                       Posted June 24, 2014.   
2                                          Posted February 3.   
3                                    Posted October 14, 2013.   
4  3 people found this review funny    Posted April 15, 2014.   

  reviews_last_edited reviews_item_id  \
0                                1250   
1                              251610   
2                              248820   
3                              250320   
4                              211420   

                                   reviews_helpful  reviews_recommend  \
0                                   No ratings yet               True   
1  15 of 20 people (75%) found this review helpful               True   
2                                   No ratings yet               True   
3   2 of 2 people (100%) found this review helpful               True   
4  35 of 43 people (81%) found this review helpful               True   

                                      reviews_review  
0  Simple yet with great replayability. In my opi...  
1  I know what you think when you see this title ...  
2  A suitably punishing roguelike platformer.  Wi...  
3  This game... is so fun. The fight sequences ha...  
4                                            Git gud

Tras este proceso, podemos ver que la columna de "user_id" y "user_url" se perdió nuevamente, por lo que volveremos a concatenarlos.

In [38]:
df_user_reviews = pd.concat([df_desanidado[["user_id", "user_url"]], df_user_reviews], axis=1)
df_user_reviews.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

                      reviews_funny            reviews_posted  \
0                                    Posted November 5, 2011.   
1                                       Posted June 24, 2014.   
2                                          Posted February 3.   
3                                    Posted October 14, 2013.   
4  3 people found this review funny    Posted April 15, 2014.   

  reviews_last_edited reviews_item_id  \
0                                1250   
1                              251610   
2                              248820   
3                              250320   
4                              211420   

                                   reviews_helpful  reviews_recommend  \
0                                   No ratings yet               True   
1  15 of 20 people (75%) found this review helpful               True   
2                                   No ratings yet               True   
3   2 of 2 people (100%) found this review helpful               True   
4  35 of 43 people (81%) found this review helpful               True   

                                      reviews_review  
0  Simple yet with great replayability. In my opi...  
1  I know what you think when you see this title ...  
2  A suitably punishing roguelike platformer.  Wi...  
3  This game... is so fun. The fight sequences ha...  
4                                            Git gud

Podemos visualizar que hay valores faltantes en algunas columnas, pero no están como nulos, probablemente porque deben tener un espacio. Lo comprobaremos y si es así, los rellenaremos como valores nulos.

In [39]:
df_user_reviews["reviews_last_edited"][0]

''

In [45]:
df_user_reviews.replace("", None, inplace=True)
df_user_reviews

user_id                                  user_url reviews_posted  \
85586       --000--      http://steamcommunity.com/id/--000--     2014-02-18   
8189        --000--      http://steamcommunity.com/id/--000--     2015-01-06   
59787       --000--      http://steamcommunity.com/id/--000--     2015-01-01   
33988       --000--      http://steamcommunity.com/id/--000--            NaT   
24925       --ace--      http://steamcommunity.com/id/--ace--            NaT   
...             ...                                       ...            ...   
197707  zzoptimuszz  http://steamcommunity.com/id/zzoptimuszz     2014-11-13   
223506  zzoptimuszz  http://steamcommunity.com/id/zzoptimuszz     2014-11-13   
68712   zzoptimuszz  http://steamcommunity.com/id/zzoptimuszz     2015-03-16   
94511   zzoptimuszz  http://steamcommunity.com/id/zzoptimuszz     2015-06-03   
120310  zzoptimuszz  http://steamcommunity.com/id/zzoptimuszz     2015-02-06   

       reviews_item_id                                 reviews_helpful  \
85586           218230                                  No ratings yet   
8189            233610  1 of 1 people (100%) found this review helpful   
59787           210770                                  No ratings yet   
33988           208650   3 of 5 people (60%) found this review helpful   
24925              730                                  No ratings yet   
...                ...                                             ...   
197707             320                                  No ratings yet   
223506             420                                  No ratings yet   
68712           230410    0 of 5 people (0%) found this review helpful   
94511           333930                                  No ratings yet   
120310             550                                  No ratings yet   

        reviews_recommend                                     reviews_review  \
85586                True  This game is just pure awesome.  One of the be...   
8189                 True  Distance is a wonderful racing game.  The seco...   
59787                True  Sanctum 2 is one really the only FPS/Tower Def...   
33988               False  I have played this game for an extensive amoun...   
24925                True  One of the best FPS competetive games i hae pl...   
...                   ...                                                ...   
197707               True  Another muliplayer of Half-Life 2, but killing...   
223506               True  Half-Life : Welp, This is way too old at first...   
68712               False  You have to log in 2 times1. Steam Login2. Gam...   
94511                True  The developers made this game awesome, made it...   
120310               True  This game is amazing! You can play with your f...   

        year  
85586   2014  
8189    2015  
59787   2015  
33988   <NA>  
24925   <NA>  
...      ...  
197707  2014  
223506  2014  
68712   2015  
94511   2015  
120310  2015  

[57397 rows x 8 columns]

Analizamos el tipo de datos y los nulos que quedaron luego de desanidar la columna "reviews".

In [41]:
tools.ver_tipo_datos(df_user_reviews)

nombre_campo                           tipo_datos  no_nulos_%  \
0              user_id                      [<class 'str'>]      100.00   
1             user_url                      [<class 'str'>]      100.00   
2        reviews_funny  [<class 'NoneType'>, <class 'str'>]       13.76   
3       reviews_posted                      [<class 'str'>]      100.00   
4  reviews_last_edited  [<class 'NoneType'>, <class 'str'>]       10.28   
5      reviews_item_id                      [<class 'str'>]      100.00   
6      reviews_helpful                      [<class 'str'>]      100.00   
7    reviews_recommend                     [<class 'bool'>]      100.00   
8       reviews_review  [<class 'str'>, <class 'NoneType'>]       99.95   

   nulos_%  nulos  
0     0.00      0  
1     0.00      0  
2    86.24  49498  
3     0.00      0  
4    89.72  51499  
5     0.00      0  
6     0.00      0  
7     0.00      0  
8     0.05     30

Hay un 86 y 89% de faltantes de datos en las columnas "reviews_funny" y "reviews_last_edited" respectivamente, por lo que decidimos eliminar estas columnas.

In [42]:
df_user_reviews = df_user_reviews.drop(columns=["reviews_funny", "reviews_last_edited"])
df_user_reviews.columns

Index(['user_id', 'user_url', 'reviews_posted', 'reviews_item_id',
       'reviews_helpful', 'reviews_recommend', 'reviews_review'],
      dtype='object')

#### Columna "reviews_posted"
La fecha se encuentra en este formato Posted April 17, 2015, pero necesitamos que la fecha esté en formato YYYY-MM-DD. Por lo tanto, procesaremos la fecha y extraeremos los elementos relevantes.

In [50]:

# Nos aseguramos de que "posted" sea de tipo datetime
df_user_reviews["reviews_posted"] = pd.to_datetime(df_user_reviews["reviews_posted"].astype(str).str.replace(r'Posted |,|\.', '', regex=True), errors='coerce')

# Creamos la columna "year" a partir de "posted"
df_user_reviews["year"] = df_user_reviews["reviews_posted"].dt.year.astype("Int64")

# Ordenamos el DataFrame por "item_id" y "year" para asegurar que la interpolación se haga correctamente
df_user_reviews = df_user_reviews.sort_values(["user_id", "year"])


df_user_reviews





user_id                                  user_url reviews_posted  \
85586       --000--      http://steamcommunity.com/id/--000--     2014-02-18   
8189        --000--      http://steamcommunity.com/id/--000--     2015-01-06   
59787       --000--      http://steamcommunity.com/id/--000--     2015-01-01   
33988       --000--      http://steamcommunity.com/id/--000--            NaT   
24925       --ace--      http://steamcommunity.com/id/--ace--            NaT   
...             ...                                       ...            ...   
197707  zzoptimuszz  http://steamcommunity.com/id/zzoptimuszz     2014-11-13   
223506  zzoptimuszz  http://steamcommunity.com/id/zzoptimuszz     2014-11-13   
68712   zzoptimuszz  http://steamcommunity.com/id/zzoptimuszz     2015-03-16   
94511   zzoptimuszz  http://steamcommunity.com/id/zzoptimuszz     2015-06-03   
120310  zzoptimuszz  http://steamcommunity.com/id/zzoptimuszz     2015-02-06   

       reviews_item_id                                 reviews_helpful  \
85586           218230                                  No ratings yet   
8189            233610  1 of 1 people (100%) found this review helpful   
59787           210770                                  No ratings yet   
33988           208650   3 of 5 people (60%) found this review helpful   
24925              730                                  No ratings yet   
...                ...                                             ...   
197707             320                                  No ratings yet   
223506             420                                  No ratings yet   
68712           230410    0 of 5 people (0%) found this review helpful   
94511           333930                                  No ratings yet   
120310             550                                  No ratings yet   

        reviews_recommend                                     reviews_review  \
85586                True  This game is just pure awesome.  One of the be...   
8189                 True  Distance is a wonderful racing game.  The seco...   
59787                True  Sanctum 2 is one really the only FPS/Tower Def...   
33988               False  I have played this game for an extensive amoun...   
24925                True  One of the best FPS competetive games i hae pl...   
...                   ...                                                ...   
197707               True  Another muliplayer of Half-Life 2, but killing...   
223506               True  Half-Life : Welp, This is way too old at first...   
68712               False  You have to log in 2 times1. Steam Login2. Gam...   
94511                True  The developers made this game awesome, made it...   
120310               True  This game is amazing! You can play with your f...   

        year  
85586   2014  
8189    2015  
59787   2015  
33988   <NA>  
24925   <NA>  
...      ...  
197707  2014  
223506  2014  
68712   2015  
94511   2015  
120310  2015  

[57397 rows x 8 columns]

In [51]:
# Rellenamos valores nulos en "year" mediante interpolación lineal por grupo (item_id)
df_user_reviews['year'] = df_user_reviews.groupby('reviews_item_id', group_keys=False)['year'].apply(lambda group: group.interpolate(method='pad') if group.notna().any() else group)

# Si aún observamos valores nulos después de la interpolación, los rellenamos con la mediana. 
df_user_reviews['year'] = df_user_reviews['year'].fillna(df_user_reviews['year'].median())

df_user_reviews

user_id                                  user_url reviews_posted  \
85586       --000--      http://steamcommunity.com/id/--000--     2014-02-18   
8189        --000--      http://steamcommunity.com/id/--000--     2015-01-06   
59787       --000--      http://steamcommunity.com/id/--000--     2015-01-01   
33988       --000--      http://steamcommunity.com/id/--000--            NaT   
24925       --ace--      http://steamcommunity.com/id/--ace--            NaT   
...             ...                                       ...            ...   
197707  zzoptimuszz  http://steamcommunity.com/id/zzoptimuszz     2014-11-13   
223506  zzoptimuszz  http://steamcommunity.com/id/zzoptimuszz     2014-11-13   
68712   zzoptimuszz  http://steamcommunity.com/id/zzoptimuszz     2015-03-16   
94511   zzoptimuszz  http://steamcommunity.com/id/zzoptimuszz     2015-06-03   
120310  zzoptimuszz  http://steamcommunity.com/id/zzoptimuszz     2015-02-06   

       reviews_item_id                                 reviews_helpful  \
85586           218230                                  No ratings yet   
8189            233610  1 of 1 people (100%) found this review helpful   
59787           210770                                  No ratings yet   
33988           208650   3 of 5 people (60%) found this review helpful   
24925              730                                  No ratings yet   
...                ...                                             ...   
197707             320                                  No ratings yet   
223506             420                                  No ratings yet   
68712           230410    0 of 5 people (0%) found this review helpful   
94511           333930                                  No ratings yet   
120310             550                                  No ratings yet   

        reviews_recommend                                     reviews_review  \
85586                True  This game is just pure awesome.  One of the be...   
8189                 True  Distance is a wonderful racing game.  The seco...   
59787                True  Sanctum 2 is one really the only FPS/Tower Def...   
33988               False  I have played this game for an extensive amoun...   
24925                True  One of the best FPS competetive games i hae pl...   
...                   ...                                                ...   
197707               True  Another muliplayer of Half-Life 2, but killing...   
223506               True  Half-Life : Welp, This is way too old at first...   
68712               False  You have to log in 2 times1. Steam Login2. Gam...   
94511                True  The developers made this game awesome, made it...   
120310               True  This game is amazing! You can play with your f...   

        year  
85586   2014  
8189    2015  
59787   2015  
33988   2014  
24925   2014  
...      ...  
197707  2014  
223506  2014  
68712   2015  
94511   2015  
120310  2015  

[57397 rows x 8 columns]

Eliminaremos la columna "reviews_posted" porque no aporta mayor información, ya que solo indica el día y el mes del posteo.

In [54]:
df_user_reviews = df_user_reviews.drop(["user_url", "reviews_posted"], axis=1)
df_user_reviews


user_id reviews_item_id  \
85586       --000--          218230   
8189        --000--          233610   
59787       --000--          210770   
33988       --000--          208650   
24925       --ace--             730   
...             ...             ...   
197707  zzoptimuszz             320   
223506  zzoptimuszz             420   
68712   zzoptimuszz          230410   
94511   zzoptimuszz          333930   
120310  zzoptimuszz             550   

                                       reviews_helpful  reviews_recommend  \
85586                                   No ratings yet               True   
8189    1 of 1 people (100%) found this review helpful               True   
59787                                   No ratings yet               True   
33988    3 of 5 people (60%) found this review helpful              False   
24925                                   No ratings yet               True   
...                                                ...                ...   
197707                                  No ratings yet               True   
223506                                  No ratings yet               True   
68712     0 of 5 people (0%) found this review helpful              False   
94511                                   No ratings yet               True   
120310                                  No ratings yet               True   

                                           reviews_review  year  
85586   This game is just pure awesome.  One of the be...  2014  
8189    Distance is a wonderful racing game.  The seco...  2015  
59787   Sanctum 2 is one really the only FPS/Tower Def...  2015  
33988   I have played this game for an extensive amoun...  2014  
24925   One of the best FPS competetive games i hae pl...  2014  
...                                                   ...   ...  
197707  Another muliplayer of Half-Life 2, but killing...  2014  
223506  Half-Life : Welp, This is way too old at first...  2014  
68712   You have to log in 2 times1. Steam Login2. Gam...  2015  
94511   The developers made this game awesome, made it...  2015  
120310  This game is amazing! You can play with your f...  2015  

[57397 rows x 6 columns]

#### Columna "reviews_review"
Esta columna tiene algunos valores nulos.

In [55]:
df_user_reviews = df_user_reviews.dropna(subset=["reviews_review"])
# Verificamos nulos
tools.ver_tipo_datos(df_user_reviews)

nombre_campo        tipo_datos  no_nulos_%  nulos_%  nulos
0            user_id   [<class 'str'>]       100.0      0.0      0
1    reviews_item_id   [<class 'str'>]       100.0      0.0      0
2    reviews_helpful   [<class 'str'>]       100.0      0.0      0
3  reviews_recommend  [<class 'bool'>]       100.0      0.0      0
4     reviews_review   [<class 'str'>]       100.0      0.0      0
5               year   [<class 'int'>]       100.0      0.0      0

In [56]:
df_user_reviews.head()

user_id reviews_item_id  \
85586  --000--          218230   
8189   --000--          233610   
59787  --000--          210770   
33988  --000--          208650   
24925  --ace--             730   

                                      reviews_helpful  reviews_recommend  \
85586                                  No ratings yet               True   
8189   1 of 1 people (100%) found this review helpful               True   
59787                                  No ratings yet               True   
33988   3 of 5 people (60%) found this review helpful              False   
24925                                  No ratings yet               True   

                                          reviews_review  year  
85586  This game is just pure awesome.  One of the be...  2014  
8189   Distance is a wonderful racing game.  The seco...  2015  
59787  Sanctum 2 is one really the only FPS/Tower Def...  2015  
33988  I have played this game for an extensive amoun...  2014  
24925  One of the best FPS competetive games i hae pl...  2014

### Carga del dataset australian_user_reviews

In [57]:
df_user_reviews.to_csv("user_reviews_cleaned.csv", index=False, encoding= "utf-8")
print(f"Se guardó el archivo")

Se guardó el archivo
